certbot/ocsp.py

```diff
@@ -65,33 +88,128 @@ def ocsp_revoked(self, cert_path, chain_path):
+        logger.error("Invalid OCSP response status for %s: %s",
+                     cert_path, response_ocsp.response_status)
+        return False
+
+    # Check OCSP signature
+    try:
+        _check_ocsp_response(response_ocsp, request, issuer)
+    except UnsupportedAlgorithm as e:
+        logger.error(str(e))
+        return False
+    except InvalidSignature:
+        logger.error('Invalid signature on OCSP response for %s', cert_path)
+        return False
+    except AssertionError as error:
+        logger.error('Invalid OCSP response for %s: %s.', cert_path, str(error))
+        return False
+
+    # Check OCSP certificate status
+    logger.debug("OCSP certificate status for %s is: %s",
+                 cert_path, response_ocsp.certificate_status)
+    return response_ocsp.certificate_status == ocsp.OCSPCertStatus.REVOKED
+
+
+def _check_ocsp_response(response_ocsp, request_ocsp, issuer_cert):
+    """Verify that the OCSP is valid for serveral criterias"""
+    # Assert OCSP response corresponds to the certificate we are talking about
+    if response_ocsp.serial_number != request_ocsp.serial_number:
+        raise AssertionError('the certificate in response does not correspond '
+                             'to the certificate in request')
+
+    # Assert signature is valid
+    _check_ocsp_response_signature(response_ocsp, issuer_cert)
+
+    # Assert issuer in response is the expected one
+    if response_ocsp.issuer_name_hash != request_ocsp.issuer_name_hash:
+        raise AssertionError('the issuer does not correspond to issuer of the certificate.')
+
+    # Assert nextUpdate is in the future, and that thisUpdate is not too old
+    if response_ocsp.next_update:
+        if response_ocsp.next_update < datetime.now():
+            raise AssertionError('next update is in the past.')
+        interval = response_ocsp.next_update - response_ocsp.this_update
+        if datetime.now() - response_ocsp.this_update > interval:
+            raise AssertionError('this update is too old.')
+
+
+def _check_ocsp_response_signature(response_ocsp, issuer_cert):
+    """Verify an OCSP response signature against certificate issuer"""
+    try:
+        # TODO: (adferrand 2019-11-01) Following line can be improved using a direct call to
+        #  response_ocsp.signature_hash_algorithm once cryptography 2.5 is released
+        #  (watch out for retro-compatibility with 2.4 though).
+        #  See https://github.com/pyca/cryptography/issues/4680
+        chosen_hash = x509._SIG_OIDS_TO_HASH[response_ocsp.signature_algorithm_oid]  # pylint: disable=protected-access
+    except KeyError:
+        raise UnsupportedAlgorithm(
+            "Signature algorithm OID:{0} not recognized"
+            .format(response_ocsp.signature_algorithm_oid)
+        )
+
+    issuer_cert.public_key().verify(
+        response_ocsp.signature,
+        response_ocsp.tbs_response_bytes,
+        padding.PKCS1v15(),
```

Comment:
If the CA is using an ECDSA intermediate cert, I think this code will have to be changed a bit. See [crypto_util.py](https://github.com/certbot/certbot/blob/693bac4b4943898ced3397fa0d6e436ef2343159/certbot/crypto_util.py#L243) or [cryptography's test_ocsp.py](https://github.com/pyca/cryptography/blob/master/tests/x509/test_ocsp.py#L384). To avoid problems in the future, can we make sure this code works with both approaches now?

If it makes sense, we should try to break out and reuse the code in `crypto_util.py`.

Reply:
Definitively! I refactored the code to be able to reuse `crypto_util`.